### PR TITLE
Support passing instance directly when invoking ViewComponents with single parameter

### PR DESCRIPTION
**Pull request by ajaybhargavb**
#5166 

@pranavkm

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewComponents/DefaultViewComponentHelper.cs [hunk 1]
```diff
@@ -134,15 +136,25 @@ private ViewComponentDescriptor SelectComponent(Type componentType)
             ViewComponentDescriptor descriptor,
             object arguments)
         {
+            IDictionary<string, object> argumentDictionary;
+
+            var methodParams = descriptor.MethodInfo.GetParameters();
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewComponents/DefaultViewComponentHelper.cs]: Seeing `GetParameters()` in this code is a red flag. I don't think we could avoid the `IsAssignableFrom` though :cry: 

should this be one the descriptor?

**Comment by ajaybhargavb** [comment]
:up: :date:

**Comment by rynowak** [comment]
Looks good, needs tests

**Comment by ajaybhargavb** [comment]
:up: :date: 
Modified some existing tests to use this enhancement. Not sure if more tests are needed.

**Comment by pranavkm** [comment]
Could we add unit tests for `InvokeCoreAsync` by making it internal?

**Comment by ajaybhargavb** [comment]
:up: :date:

File: test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/ViewComponents/DefaultViewComponentHelperTest.cs [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+                invokerFactory,
+                viewBufferScope);
+
+            var descriptor = CreateDescriptorForType(typeof(TestViewComponent));
+
+            // Act
+            var argumentDictionary = helper.GetArgumentDictionary(descriptor, new { a = 0 });
+
+            // Assert
+            Assert.Equal(1, argumentDictionary.Count);
```
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/ViewComponents/DefaultViewComponentHelperTest.cs]: `Assert.Collection` takes care of the Count requirements. You can drop this.

File: test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/ViewComponents/DefaultViewComponentHelperTest.cs [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+                });
+        }
+
+        private ViewComponentDescriptor CreateDescriptorForType(Type componentType)
+        {
+            var provider = CreateProvider(componentType);
+            return provider.GetViewComponents().First();
+        }
+
+        private class TestViewComponent
```
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/ViewComponents/DefaultViewComponentHelperTest.cs]: Could we have ViewComponents with multiple params, with `object` param type and one with `IDictionary<string, object>`. The last two are special cases in the `IsAssignableFrom` \ `PropertyHelper` scenarios. The first one covers the alternate code path in `GetArgumentDictionary`

**Comment by pranavkm** [comment]
:watch: for a few more tests.

File: src/Microsoft.AspNetCore.Mvc.Abstractions/Abstractions/ActionDescriptor.cs [hunk 4]
```diff
@@ -36,13 +36,19 @@ public ActionDescriptor()
         /// </summary>
         public IList<IActionConstraintMetadata> ActionConstraints { get; set; }
 
+        /// <summary>
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Abstractions/Abstractions/ActionDescriptor.cs]: Lets get this in a separate commit.

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewComponents/DefaultViewComponentHelper.cs [hunk 5]
```diff
@@ -130,19 +132,28 @@ private ViewComponentDescriptor SelectComponent(Type componentType)
...
 
-        private async Task<IHtmlContent> InvokeCoreAsync(
-            ViewComponentDescriptor descriptor,
-            object arguments)
+        // Internal for testing
+        internal IDictionary<string, object> GetArgumentDictionary(ViewComponentDescriptor descriptor, object arguments)
         {
+            if (descriptor.Parameters.Count == 1 && descriptor.Parameters[0].ParameterType.IsAssignableFrom(arguments.GetType()))
+            {
+                return new Dictionary<string, object>(StringComparer.OrdinalIgnoreCase)
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewComponents/DefaultViewComponentHelper.cs]: Ordinal~~IgnoreCase~~

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewComponents/ViewComponentDescriptor.cs [hunk 6]
```diff
@@ -124,5 +126,10 @@ public string DisplayName
         /// Gets or sets the <see cref="System.Reflection.MethodInfo"/> to invoke.
         /// </summary>
         public MethodInfo MethodInfo { get; set; }
+
+        /// <summary>
+        /// Gets or sets the parameters associated with the method described by <see cref="MethodInfo"/>.
+        /// </summary>
+        public IList<ParameterInfo> Parameters { get; set; }
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewComponents/ViewComponentDescriptor.cs]: `IEnumerable` \ `IReadOnlyList`?

File: src/Microsoft.AspNetCore.Mvc.Abstractions/Abstractions/ActionDescriptor.cs (hunk 4 above)
Comment (ajaybhargavb) [inline src/Microsoft.AspNetCore.Mvc.Abstractions/Abstractions/ActionDescriptor.cs]: Will put this in a different commit when I squash before I merge.

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewComponents/DefaultViewComponentHelper.cs (hunk 5 above)
Comment (ajaybhargavb) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewComponents/DefaultViewComponentHelper.cs]: Spoke offline. Not doing this because `PropertyHelper.ObjectToDictionary()` also uses `OrdinalIgnoreCase`.

**Comment by ajaybhargavb** [comment]
:up: :date:

File: test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/ViewComponents/DefaultViewComponentHelperTest.cs [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+        {
+            // Arrange
+            var helper = CreateHelper();
+            var descriptor = CreateDescriptorForType(typeof(ViewComponentSingleParam));
+
+            // Act
+            var argumentDictionary = helper.GetArgumentDictionary(descriptor, 0);
+
+            // Assert
+            Assert.Equal(1, argumentDictionary.Count);
```
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/ViewComponents/DefaultViewComponentHelperTest.cs]: This can be removed.

File: test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/ViewComponents/DefaultViewComponentHelperTest.cs [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+            // Arrange
+            var helper = CreateHelper();
+            var descriptor = CreateDescriptorForType(typeof(ViewComponentSingleParam));
+            var expectedValue = new object();
+
+            // Act
+            var argumentDictionary = helper.GetArgumentDictionary(descriptor, expectedValue);
+
+            // Assert
+            Assert.Equal(1, argumentDictionary.Count);
```
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/ViewComponents/DefaultViewComponentHelperTest.cs]: This can be removed.

File: test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/ViewComponents/DefaultViewComponentHelperTest.cs [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+            Assert.Collection(argumentDictionary,
+                item =>
+                {
+                    Assert.Equal("o", item.Key);
+                    Assert.IsType(typeof(object), item.Value);
+                    Assert.Equal(expectedValue, item.Value);
```
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/ViewComponents/DefaultViewComponentHelperTest.cs]: `Assert.Same`

**Comment by pranavkm** [comment]
:shipit:
